File: Algorithms/0665_Non-decreasing_Array/Python/Non-decreasing_Array_Solution_2.py

# Space: O(1)
# Time: O(n)

class Solution:
    def checkPossibility(self, nums):
        length = len(nums)
        if length <= 1: return True
        slow, fast = -1, 0

        while fast < length - 1:
            if nums[fast] > nums[fast + 1]:
                if slow != -1: return False
                slow = fast
            fast += 1

        if slow in [-1, 0, length - 2]: return True
        if nums[slow - 1] <= nums[slow + 1]: return True
        if nums[slow] <= nums[slow + 2]: return True
        return False

File: Algorithms/0665_Non-decreasing_Array/Java/Non-DecreasingArraySolution2.java

// Space: O(n)
// Time: O(n)

class Solution {
    public boolean checkPossibility(int[] nums) {
        int length = nums.length;
        if (length <= 1) return true;
        int slow = -1, fast = 0;
        while (fast < length - 1) {
            if (!(nums[fast] <= nums[fast + 1])) {
                if (slow != -1) return false;
                slow = fast;
            }
            fast++;
        }
        if (slow == -1 || slow == 0 || slow == length - 2) return true;
        if (nums[slow - 1] <= nums[slow + 1]) return true;
        if (nums[slow] <= nums[slow + 2]) return true;
        return false;
    }
}




